lib/src/Pos/Diffusion/Full/Block.hs

```diff
@@ -335,7 +336,20 @@ streamBlocks logTrace smM logic streamWindow enqueue nodeId tipHeader checkpoint
                          processBlocks batchSize n' (block : blocks) blockChan
 
     writeStreamEnd :: Conc.TBQueue StreamEntry -> IO ()
-    writeStreamEnd blockChan = atomically $ Conc.writeTBQueue blockChan StreamEnd
+    writeStreamEnd blockChan = writeBlock 1024 blockChan StreamEnd
+
+    writeBlock :: Int -> Conc.TBQueue StreamEntry -> StreamEntry -> IO ()
+    writeBlock delay _ _ | delay >= 4000000 = do
```

Comment:
This is a little opaque. Why 4000000?

Reply:
4000000 corresponds to a 4 seconds delay, we need a limit for when we give up, and time time we sleep doubles after every time we try so a 4 second limit seemed reasonable.